packaging/tags.py

```diff
@@ -420,16 +420,80 @@ def _have_compatible_glibc(required_major, minimum_minor):
     return _check_glibc_version(version_str, required_major, minimum_minor)
 
 
+def _get_elf_header():
+    # type: () -> Optional[List[int]]
+    try:
+        with open(sys.executable, "rb") as f:
+            elf_header_raw = f.read(40)  # read 40 first bytes of ELF header
+    except (IOError, OSError, TypeError):
+        return None
+    if elf_header_raw is None or len(elf_header_raw) < 40:
+        return None
+    if isinstance(elf_header_raw, str):  # pragma: no cover
+        # python 2.7
+        elf_header = [ord(c) for c in elf_header_raw]
+    else:  # pragma: no cover
+        # python 3.x
+        elf_header = [b for b in elf_header_raw]
+    if elf_header[0:4] == [0x7F, 0x45, 0x4C, 0x46]:  # ELF magic number
+        return elf_header
+    return None
+
+
+def _is_linux_armhf():
+    # type: () -> bool
+    # hard-float ABI can be detected from the ELF header of the running
+    # process
+    elf_header = _get_elf_header()
+    if elf_header is None:
+        return False
+    result = elf_header[4:5] == [1]  # 32-bit ELF
+    result &= elf_header[5:6] == [1]  # little-endian
+    result &= elf_header[18:20] == [0x28, 0]  # ARM machine
+    result &= elf_header[39:40] == [5]  # ARM EABIv5
+    result &= (elf_header[37:38][0] & 4) == 4  # EF_ARM_ABI_FLOAT_HARD
```

Comment:
I think it is harder than it needs to be to compare this to the spec. Can we try using something like `struct.unpack_from` instead? e.g.

```
flags = struct.unpack_from('<I', elf_header, offset=0x24)
result &= (flags & 0xFF000000) == 0x05000000
result &= (flags & 0x00000400) == 0x00000400
```

Reply:
Reworked implementation.